ArchiSteamFarm/Bot.cs

```diff
@@ -2836,6 +2837,136 @@ public sealed class Bot : IAsyncDisposable {
 			}
 		}
 
+		private void OnInventoryChanged() {
+			Utilities.InBackground(CardsFarmer.OnNewItemsNotification);
+
+			if (BotConfig.BotBehaviour.HasFlag(BotConfig.EBotBehaviour.DismissInventoryNotifications)) {
+				Utilities.InBackground(ArchiWebHandler.MarkInventory);
+			}
+
+			if (BotConfig.SendSetsOnCompleted) {
+				Utilities.InBackground(SendCompletedSets);
+			}
+		}
+
+		private async Task SendCompletedSets() {
+			IDocument? badgePage = await ArchiWebHandler.GetBadgePage(1).ConfigureAwait(false);
+
+			if (badgePage == null) {
+				ArchiLogger.LogGenericWarning(Strings.WarningCouldNotCheckBadges);
+
+				return;
+			}
+
+			List<IElement> craftButtons = badgePage.SelectNodes("//a[@class='badge_craft_button']");
+
+			if (craftButtons.Count == 0) {
+				// no craftable badges
+				return;
+			}
+
+			HashSet<ulong> appIds = new HashSet<ulong>(craftButtons.Count);
+
+			foreach (string? badgeUri in craftButtons.Select(htmlNode => htmlNode.GetAttribute("href"))) {
+				if (string.IsNullOrEmpty(badgeUri)) {
+					ArchiLogger.LogNullError(nameof(badgeUri));
+
+					continue;
+				}
+
+				const string uriSegmentBeforeAppId = "/gamecards/";
+				int index = badgeUri.LastIndexOf(uriSegmentBeforeAppId, StringComparison.Ordinal);
+
+				if (index < 0) {
+					ArchiLogger.LogNullError(nameof(index));
+
+					continue;
+				}
+
+				index += uriSegmentBeforeAppId.Length;
+
+				if (index >= badgeUri.Length) {
+					ArchiLogger.LogNullError(nameof(index));
+
+					continue;
+				}
+
+				string? appIdText = string.Concat(badgeUri.Skip(index).TakeWhile(char.IsDigit));
+
+				if (string.IsNullOrEmpty(appIdText) || !ulong.TryParse(appIdText, out ulong appId)) {
+					ArchiLogger.LogNullError($"{nameof(appIdText)} || {nameof(appId)}");
+
+					continue;
+				}
+
+				appIds.Add(appId);
+			}
+
+			if (appIds.Count == 0) {
+				// Could not extract appIds for craftable badges - No need to load the inventory at this point
+				return;
+			}
+
+			HashSet<Steam.Asset> inventory;
+
+			try {
+				inventory = await ArchiWebHandler.GetInventoryAsync(SteamID)
+					.Where(item => (item.Type == Steam.Asset.EType.TradingCard) || (item.Type == Steam.Asset.EType.FoilTradingCard))
+					.Where(item => appIds.Contains(item.RealAppID))
+					.ToHashSetAsync()
+					.ConfigureAwait(false);
+			} catch (HttpRequestException e) {
+				ArchiLogger.LogGenericWarningException(e);
+
+				return;
+			} catch (Exception e) {
+				ArchiLogger.LogGenericException(e);
+
+				return;
+			}
+
+			if (inventory.Count == 0) {
+				ArchiLogger.LogGenericWarning(string.Format(Strings.ErrorIsEmpty), nameof(inventory));
+
+				return;
+			}
+
+			HashSet<Steam.Asset> itemsToGive = inventory.GroupBy(item => (item.RealAppID, item.Type)).Select(
+				cardsOfAppAndType => {
+					// We still need untradable items here, as otherwise we could wrongfully assume that the set consists only of cards A, B and C due to all cards of classId D being untradable
```

Comment:
I don't like this logic. First of all, it fetches both tradable and untradable items and determines here whether the whole set is tradable. You don't know how many cards are in a set and whether our inventory didn't change between your badge page request and this one.

```c#
		internal static async Task<byte> GetCardsForSet(uint appID) {
			if (appID == 0) {
				ArchiBot.ArchiLogger.LogNullError(nameof(appID));

				return 0;
			}

			using IDocument htmlDocument = await GetGameCardsPage(appID).ConfigureAwait(false);

			if (htmlDocument == null) {
				return 0;
			}

			List<IElement> htmlNodes = htmlDocument.SelectNodes("//div[@class='badge_detail_tasks']//div[@class='badge_card_set_text ellipsis']");

			if ((htmlNodes.Count == 0) || (htmlNodes.Count % 2 != 0)) {
				return await SteamCardExchange.GetCardsForSet(appID).ConfigureAwait(false);
			}

			return (byte) (htmlNodes.Count / 2);
		}
```

Make use of this, just eradicate the call to SCE, we probably don't need logic for banned sets.

Reply:
Even with this it doesn't really get better :(